<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 3.9.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/uploads/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/uploads/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/uploads/favicon-16x16.png">
  <link rel="mask-icon" href="/uploads/favicon.png" color="#222">
  <meta name="msapplication-config" content="/uploads/browserconfig.xml">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css">
  <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox@3/dist/jquery.fancybox.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"adrrrrrrrian.github.io","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false,"b2t":false,"scrollpercent":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":true,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="Do NOT just catch Exception or Throwable, which will also catch RuntimeException">
<meta name="keywords" content="Java">
<meta property="og:type" content="article">
<meta property="og:title" content="Java Exception Handling Best Practices">
<meta property="og:url" content="https://adrrrrrrrian.github.io/2020/04/21/java-exception-handling-best-practices/index.html">
<meta property="og:site_name" content="Adrian&#39;s blog">
<meta property="og:description" content="Do NOT just catch Exception or Throwable, which will also catch RuntimeException">
<meta property="og:locale" content="en">
<meta property="og:updated_time" content="2020-05-05T15:11:31.853Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Java Exception Handling Best Practices">
<meta name="twitter:description" content="Do NOT just catch Exception or Throwable, which will also catch RuntimeException">

<link rel="canonical" href="https://adrrrrrrrian.github.io/2020/04/21/java-exception-handling-best-practices/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>Java Exception Handling Best Practices | Adrian's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<link rel="stylesheet" href="/css/prism-tomorrow.css" type="text/css">
<link rel="stylesheet" href="/css/prism-line-numbers.css" type="text/css"></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta custom-logo">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Adrian's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-fw fa-home"></i>Home</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-fw fa-tags"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-fw fa-th"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-fw fa-archive"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://adrrrrrrrian.github.io/2020/04/21/java-exception-handling-best-practices/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/uploads/favicon.png">
      <meta itemprop="name" content="adrrrrrrrian">
      <meta itemprop="description" content="I think that we can make it">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Adrian's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          Java Exception Handling Best Practices
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-04-21 06:40:45" itemprop="dateCreated datePublished" datetime="2020-04-21T06:40:45-04:00">2020-04-21</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Tech/" itemprop="url" rel="index"><span itemprop="name">Tech</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <h4 id="Do-NOT-just-catch-Exception-or-Throwable-which-will-also-catch-RuntimeException"><a href="#Do-NOT-just-catch-Exception-or-Throwable-which-will-also-catch-RuntimeException" class="headerlink" title="Do NOT just catch Exception or Throwable, which will also catch RuntimeException"></a>Do <strong>NOT</strong> just catch Exception or Throwable, which will also catch RuntimeException</h4><a id="more"></a>

<pre><code>// Bad
try {
  ...
} catch (Throwable th) {
  ...
}
// Good
try {
  ...
} catch (SpecificException ex) {
  ...
}</code></pre><h4 id="When-catching-an-exception-do-NOT-just-throw-it-Instead-do-something-like-throw-it-with-some-description-log-an-error-or-stop-the-process"><a href="#When-catching-an-exception-do-NOT-just-throw-it-Instead-do-something-like-throw-it-with-some-description-log-an-error-or-stop-the-process" class="headerlink" title="When catching an exception, do NOT just throw it. Instead, do something like throw it with some description, log an error or stop the process."></a>When catching an exception, do <strong>NOT</strong> just throw it. Instead, do something like <strong>throw it with some description</strong>, <strong>log an error</strong> or <strong>stop the process.</strong></h4><p><strong>If you catch an exception and do nothing or just log and then let the method continue, include a comment why it&#39;s ok to ignore the exception</strong></p>
<pre><code>// Bad
try {
  ...
} catch (SpecificException ex) {
  throw ex;
}
// Good 
try {
  ...
} catch (SpecificException ex) {
  log.error(&quot;This function does not work due to SpecificException: &quot;, ex);
}</code></pre><p>OR</p>
<pre><code>try {
  ...
} catch (SpecificException ex) {
  throw new SpecificException(&quot;This function does not work due to SpecificException: &quot;, ex);
}</code></pre><p>OR</p>
<pre><code>try {
  ...
} catch (SpecificException ex) {
  // Let the method continue because xyz...
  log.info(&quot;xxxxxx&quot;);
}</code></pre><h4 id="When-catching-an-exception-we-may-want-to-log-an-error-or-warning-and-do-NOT-just-log-the-string-without-the-stacktrace"><a href="#When-catching-an-exception-we-may-want-to-log-an-error-or-warning-and-do-NOT-just-log-the-string-without-the-stacktrace" class="headerlink" title="When catching an exception, we may want to log an error or warning, and do NOT just log the string without the stacktrace"></a>When catching an exception, we may want to log an error or warning, and do <strong>NOT</strong> just log the string without the stacktrace</h4><pre><code>// Bad
try {
  ...
} catch (SpecificException ex) {
  log.error(&quot;Something wrong...&quot;);
}
// Good 
try {
  ...
} catch (SpecificException ex) {
  log.error(&quot;This function does not work due to SpecificException: &quot;, ex);
}</code></pre><h4 id="How-all-kinds-of-different-methods-in-Exception-work-in-the-console"><a href="#How-all-kinds-of-different-methods-in-Exception-work-in-the-console" class="headerlink" title="How all kinds of different methods in Exception work in the console"></a>How all kinds of different methods in Exception work in the console</h4><pre><code>public class TestAllMethods {
    static final Log log = LogFactory.getLog(TestAllMethods.class);
    public static void main(String[] args) {
        try {
            System.out.println(10/0);
        } catch (ArithmeticException ex) {
            ex.printStackTrace();
            //You will get: 
            // java.lang.ArithmeticException: / by zero
            //        at TestAllMethods.main(TestAllMethods.java:10)

            log.error(&quot;Try just the exception: \n&quot;, ex);
            // You will get: 
            // Sep 06, 2017 3:45:54 PM TestAllMethods main
            // SEVERE: Try just the exception: 

            // java.lang.ArithmeticException: / by zero
            //      at TestAllMethods.main(TestAllMethods.java:10)


            log.error(&quot;Try getMessage(): \n&quot; + ex.getMessage());  
            // You will get:
            //  Sep 06, 2017 3:45:54 PM TestAllMethods main
            //  SEVERE: Try getMessage(): 
            //  / by zero

            log.error(&quot;Try toString(): \n&quot; + ex.toString());
            // You will get:
             // Sep 06, 2017 3:45:54 PM TestAllMethods main
            // SEVERE: Try toString(): 
            // java.lang.ArithmeticException: / by zero


            log.error(&quot;Try getStackTrace(): \n&quot; + ex.getStackTrace());
            // You will get:
            // Sep 06, 2017 3:45:54 PM TestAllMethods main
            // SEVERE: Try getStackTrace(): 
            // [Ljava.lang.StackTraceElement;@5a07e868


            log.error(&quot;Try Arrays.toString(ex.getStackTrace()): \n&quot; + Arrays.toString(ex.getStackTrace()));
            // You will get:
            // Sep 06, 2017 3:45:54 PM TestAllMethods main
            // SEVERE: Try Arrays.toString(ex.getStackTrace()): 
            // [TestAllMethods.main(TestAllMethods.java:10)]

            log.error(&quot;Try getCause(): \n&quot; + ex.getCause());
            // You will get: 
            // Sep 06, 2017 3:45:54 PM TestAllMethods main
            // SEVERE: Try getCause(): 
            // null
        }
    }
}</code></pre><p><strong>Conclusions:</strong></p>
<p>1) <strong>ex.toString()</strong> is better than <strong>ex.getMessage()</strong> as <strong>toString</strong> includes the exception class name, which <strong>getMessage</strong> does not;</p>
<p>2) <strong>ex.getStackTrace()</strong> will return an array of <strong>StackTraceElement</strong>, so do <strong>NOT</strong> print that directly, instead we need <strong>Arrays.toString(ex.getStackTrace())</strong> to show the content of that exception;</p>
<p>3) <strong>ex.getCause()</strong> will show you the cause of this exception or null if the cause is nonexistent or unknown.</p>
<h4 id="Try-to-avoid-duplicated-stack-trace-by-NOT-log-an-exception-with-the-stack-trace-and-then-throw-it"><a href="#Try-to-avoid-duplicated-stack-trace-by-NOT-log-an-exception-with-the-stack-trace-and-then-throw-it" class="headerlink" title="Try to avoid duplicated stack trace by NOT log an exception with the stack trace and then throw it"></a>Try to <strong>avoid</strong> duplicated stack trace by <strong>NOT</strong> log an exception with the stack trace and then throw it</h4><pre><code>// Bad
public static void main(String[] args) {
    try {
        test();
    } catch (SQLException ex) {
        log.error(&quot;second level: &quot;, ex);
    }
}
private static void test() throws SQLException {
    try {
        throw new SQLException();
    } catch (SQLException ex) {
        log.error(&quot;first level: &quot;, ex);
        throw new SQLException(ex);
    }
}
// You will get: 
Sep 06, 2017 4:12:40 PM MultipleException test
SEVERE: first level: 
java.sql.SQLException
    at MultipleException.test(MultipleException.java:19)
    at MultipleException.main(MultipleException.java:11)

Sep 06, 2017 4:12:40 PM MultipleException main
SEVERE: second level: 
java.sql.SQLException: java.sql.SQLException
    at MultipleException.test(MultipleException.java:22)
    at MultipleException.main(MultipleException.java:11)
Caused by: java.sql.SQLException
    at MultipleException.test(MultipleException.java:19)
    ... 1 more


// Good 
public static void main(String[] args) {
    try {
        test();
    } catch (SQLException ex) {
        log.error(&quot;wrong: &quot;, ex);
    }
}
private static void test() throws SQLException {
    try {
        throw new SQLException();
    } catch (SQLException ex) {
        //log.error(&quot;first level: &quot;, ex);
        throw new SQLException(&quot;first level: &quot;, ex);
    }
}
// You will get: 
Sep 06, 2017 4:16:25 PM MultipleException main
SEVERE: wrong: 
java.sql.SQLException: first level: 
    at MultipleException.test(MultipleException.java:22)
    at MultipleException.main(MultipleException.java:11)
Caused by: java.sql.SQLException
    at MultipleException.test(MultipleException.java:19)
    ... 1 more</code></pre><h4 id="Remember-to-release-the-resource-using-try-with-resource-or-try-catch-finally-block"><a href="#Remember-to-release-the-resource-using-try-with-resource-or-try-catch-finally-block" class="headerlink" title="Remember to release the resource using try-with-resource or try-catch-finally block"></a>Remember to <strong>release the resource</strong> using try-with-resource or try-catch-finally block</h4><pre><code>// Bad
try {
  Connection conn = getConnection();
} catch (SpecificException ex) {
  log.error(&quot;This function does not work due to SpecificException: &quot;, ex);
}


// Good 
try {
  Connection conn = getConnection();
  ...
} catch (SpecificException ex) {
  log.error(&quot;This function does not work due to SpecificException: &quot;, ex);
} finally {
  context.close(conn);
}</code></pre><p>OR</p>
<pre><code>try (Connection conn = getConnection()) {
  ...
} catch (SpecificException ex) {
  log.error(&quot;This function does not work due to SpecificException: &quot;, ex);
} </code></pre><h4 id="If-you-are-using-try-with-resource-the-implicit-finally-block-could-throw-an-exception-A-and-if-it-happens-when-something-in-try-block-also-throws-an-exception-B-A-will-be-suppressed-exception-of-B-and-log-warn-error-info-will-print-both-exceptions-out"><a href="#If-you-are-using-try-with-resource-the-implicit-finally-block-could-throw-an-exception-A-and-if-it-happens-when-something-in-try-block-also-throws-an-exception-B-A-will-be-suppressed-exception-of-B-and-log-warn-error-info-will-print-both-exceptions-out" class="headerlink" title="If you are using try-with-resource, the implicit finally() block could throw an exception A, and if it happens when something in try block also throws an exception B, A will be suppressed exception of B, and log.warn/error/info() will print both exceptions out."></a>If you are using try-with-resource, the implicit finally() block could throw an exception A, and if it happens when something in try block also throws an exception B, A will be suppressed exception of B, and log.warn/error/info() will print both exceptions out.</h4><figcaption>20, 2017 1:17:02 PM toml.main.Main logException</figcaption><pre class="line-numbers language-Nov"><code class="language-Nov">WARNING: logException - MyExceptionA: 
toml.main.MyExceptionA: MyExceptionA in run
    at toml.main.MyAutoClose.run(MyAutoClose.java:7)
    at toml.main.Main.callMyAutoClose(Main.java:25)
    at toml.main.Main.main(Main.java:12)
    Suppressed: toml.main.MyExceptionB: MyExceptionB in close
        at toml.main.MyAutoClose.close(MyAutoClose.java:13)
        at toml.main.Main.callMyAutoClose(Main.java:26)
        ... 1 more<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>

    </div>

    
    
    

      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/Java/" rel="tag"># Java</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2020/04/17/netty-in-action-chapter-9/" rel="prev" title="Netty in Action Chapter 9">
      <i class="fa fa-chevron-left"></i> Netty in Action Chapter 9
    </a></div>
      <div class="post-nav-item">
    <a href="/2020/04/23/work-from-home-ti-hui/" rel="next" title="work from home 体会">
      work from home 体会 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-4"><a class="nav-link" href="#Do-NOT-just-catch-Exception-or-Throwable-which-will-also-catch-RuntimeException"><span class="nav-number">1.</span> <span class="nav-text">Do NOT just catch Exception or Throwable, which will also catch RuntimeException</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#When-catching-an-exception-do-NOT-just-throw-it-Instead-do-something-like-throw-it-with-some-description-log-an-error-or-stop-the-process"><span class="nav-number">2.</span> <span class="nav-text">When catching an exception, do NOT just throw it. Instead, do something like throw it with some description, log an error or stop the process.</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#When-catching-an-exception-we-may-want-to-log-an-error-or-warning-and-do-NOT-just-log-the-string-without-the-stacktrace"><span class="nav-number">3.</span> <span class="nav-text">When catching an exception, we may want to log an error or warning, and do NOT just log the string without the stacktrace</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#How-all-kinds-of-different-methods-in-Exception-work-in-the-console"><span class="nav-number">4.</span> <span class="nav-text">How all kinds of different methods in Exception work in the console</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#Try-to-avoid-duplicated-stack-trace-by-NOT-log-an-exception-with-the-stack-trace-and-then-throw-it"><span class="nav-number">5.</span> <span class="nav-text">Try to avoid duplicated stack trace by NOT log an exception with the stack trace and then throw it</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#Remember-to-release-the-resource-using-try-with-resource-or-try-catch-finally-block"><span class="nav-number">6.</span> <span class="nav-text">Remember to release the resource using try-with-resource or try-catch-finally block</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#If-you-are-using-try-with-resource-the-implicit-finally-block-could-throw-an-exception-A-and-if-it-happens-when-something-in-try-block-also-throws-an-exception-B-A-will-be-suppressed-exception-of-B-and-log-warn-error-info-will-print-both-exceptions-out"><span class="nav-number">7.</span> <span class="nav-text">If you are using try-with-resource, the implicit finally() block could throw an exception A, and if it happens when something in try block also throws an exception B, A will be suppressed exception of B, and log.warn/error/info() will print both exceptions out.</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="adrrrrrrrian"
      src="/uploads/favicon.png">
  <p class="site-author-name" itemprop="name">adrrrrrrrian</p>
  <div class="site-description" itemprop="description">I think that we can make it</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">30</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">2</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">9</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/adrrrrrrrian" title="GitHub → https://github.com/adrrrrrrrian" rel="noopener" target="_blank"><i class="fa fa-fw fa-github"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:qihu1992@hotmail.com" title="E-Mail → mailto:qihu1992@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2020</span>
  <span class="with-love">
    <i class="fa fa-"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">adrrrrrrrian</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="//cdn.jsdelivr.net/npm/jquery@3/dist/jquery.min.js"></script>
  <script src="//cdn.jsdelivr.net/gh/fancyapps/fancybox@3/dist/jquery.fancybox.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>
<script src="/js/utils.js"></script><script src="/js/motion.js"></script>
<script src="/js/schemes/pisces.js"></script>
<script src="/js/next-boot.js"></script>



  















  

  

</body>
</html>
